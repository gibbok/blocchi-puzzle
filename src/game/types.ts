export enum DirectionEnum {
  N = 'N',
  E = 'E',
  S = 'S',
  W = 'W',
}

export const NO = DirectionEnum.N;
export const ES = DirectionEnum.E;
export const SO = DirectionEnum.S;
export const WE = DirectionEnum.W;

export enum TetroEnum {
  Z = 'Z',
  S = 'S',
  J = 'J',
  T = 'T',
  I = 'I',
  L = 'L',
  O = 'O',
}

export enum NoTetroEnum {
  NoTetro,
}
export const NoTetro = NoTetroEnum.NoTetro;

export type Tile = NoTetroEnum.NoTetro | TetroEnum;

export type TetroPieces = Record<TetroEnum, Record<DirectionEnum, Tetro>>;

export type TetroRow = readonly Tile[];

// The tetromino, a geometric shape composed of four squares/tiles
export type Tetro = readonly TetroRow[];

export type TetroDef = Readonly<{
  type: TetroEnum;
  direction: DirectionEnum;
  x: number;
  y: number;
}>;

export const Z = TetroEnum.Z;
export const S = TetroEnum.S;
export const J = TetroEnum.J;
export const T = TetroEnum.T;
export const I = TetroEnum.I;
export const L = TetroEnum.L;
export const O = TetroEnum.O;

export type Position = number;

export type BoardRow = readonly Tile[];

export type BoardMutable = Tile[][];

// This is the space on which the game is played.
// Tetriminos fall from the top of the Board, and then players turn and move them into their desired places at the bottom.
export type Board = readonly BoardRow[];

export enum ScreenEnum {
  Intro = 'Intro',
  Game = 'Game',
  Over = 'Over',
}

export type InternalState = Readonly<{
  board: Board;
  score: number; // some point based on lines done
  level: number; // based on score, every 1000 score you go to next level, speed is faster
  lines: number; // number of lines done
  currentTetro: TetroDef;
  nextTetro: TetroDef;
  isPlay: boolean;
  isGameOver: boolean;
  screen: ScreenEnum;
}>;

export type PublicState = Omit<InternalState, 'currentTetro' | 'isPlay' | 'isGameOver'>;

export enum KeyEnum {
  Esc = 27,
  Space = 32,
  Left = 37,
  Up = 38,
  Right = 39,
  Down = 40,
}

export type CallBack = () => void;

export type Action = Readonly<{
  type: string;
}>;

export type TransitionState = 'entering' | 'entered' | 'exiting' | 'exited';
export interface TransitionStyles {
  entering?: React.CSSProperties;
  entered?: React.CSSProperties;
  exiting?: React.CSSProperties;
  exited?: React.CSSProperties;
}
