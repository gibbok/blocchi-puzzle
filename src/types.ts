export enum DirectionEnum {
  N = 'N',
  E = 'E',
  S = 'S',
  W = 'W'
}

export const NO = DirectionEnum.N;
export const ES = DirectionEnum.E;
export const SO = DirectionEnum.S;
export const WE = DirectionEnum.W;

export enum TetroEnum {
  Z = 'Z',
  S = 'S',
  J = 'J',
  T = 'T',
  I = 'I',
  O = 'O'
}

export type NoTetro = 0;
export type Block = NoTetro | TetroEnum;

export type TetroPieces = Record<TetroEnum, Record<DirectionEnum, Tetro>>;

export type TetroRow = readonly Block[];

export type Tetro = readonly (TetroRow)[];

export type TetroDef = Readonly<{
  type: TetroEnum;
  direction: DirectionEnum;
  x: number;
  y: number;
}>;

export const Z = TetroEnum.Z;
export const S = TetroEnum.S;
export const J = TetroEnum.J;
export const T = TetroEnum.T;
export const I = TetroEnum.I;
export const O = TetroEnum.O;

export type Position = number;

export type BoardRow = readonly Block[];

export type Board = readonly (BoardRow)[];

export type InternalState = Readonly<{
  board: Board;
  score: number;
  level: number;
  lines: number;
  currentTetro: TetroDef;
  nextTetro: TetroDef;
  isPlay: boolean;
}>;

export type PubicState = Omit<InternalState, 'currentTetro' | 'isPlay'>;

export enum KeyEnum {
  Esc = 27,
  Space = 32,
  Left = 37,
  Up = 38,
  Right = 39,
  Down = 40
}

export type CallBack = () => void;

export type Action = Readonly<{
  type: string;
}>;

export enum ActionEnum {
  MoveDown = 'MoveDown',
  MoveRight = 'MoveRight',
  MoveLeft = 'MoveLeft',
  MoveUp = 'MoveUp'
}
