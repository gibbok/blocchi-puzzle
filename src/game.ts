export const calcScore = (rowCompleted: number) => rowCompleted * 100;

export const calcLevel = (score: number) => Math.ceil((score + 1) / 1000);
