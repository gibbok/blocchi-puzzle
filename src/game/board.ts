import { Board, Tile, TetroEnum, DirectionEnum, NoTetro } from './types';
import { getTetroFromPieces, occupied } from './tetromino';
import { pipe } from 'fp-ts/lib/pipeable';
import { BOARD_CELLS } from './settings';

export const mkEmptyBoard = (rows: number, columns: number): Board =>
  [...Array(rows)].fill([...Array(columns).fill(NoTetro)]);

// TODO add test
export const copyBoard = (b: Board) => b.map(r => r.map(c => c));

export const addTetroToBoard = (t: TetroEnum, d: DirectionEnum, x: number, y: number, b: Board) => {
  const tetro = getTetroFromPieces(t, d);
  const bn = copyBoard(b); // copy board
  tetro.forEach((tR, tRx) =>
    tR.forEach(
      (tC, tCx) =>
        (bn[tRx + y][tCx + x] = tetro[tRx][tCx] !== 0 ? tetro[tRx][tCx] : bn[tRx + y][tCx + x])
    )
  );
  return bn;
};

export const recFindAvailablePos = (
  type: TetroEnum,
  d: DirectionEnum,
  x: number,
  y: number,
  b: Board,
  towardsX: number,
  towardsY: number
): number => {
  const isOccupied = occupied(type, d, x, y, b);
  return isOccupied
    ? recFindAvailablePos(type, d, x - towardsX, y - towardsY, b, towardsX, towardsY)
    : y !== 0
    ? y
    : x;
};

export const recFindAvailablePosX = (
  type: TetroEnum,
  d: DirectionEnum,
  x: number,
  y: number,
  b: Board,
  towardsX: number
): number => recFindAvailablePos(type, d, x, y, b, towardsX, 0);

export const recFindAvailablePosY = (
  type: TetroEnum,
  d: DirectionEnum,
  x: number,
  y: number,
  b: Board,
  towardsY: number
): number => recFindAvailablePos(type, d, x, y, b, 0, towardsY);

export const getCompleteRowIdxs = (b: Board): number[] =>
  b.flatMap((row, idx) => (row.every(cell => cell !== NoTetro) ? [idx] : []));

export const removeCompleteRowFromBoard = (
  b: Board,
  lineIdxs: number[]
): Readonly<{
  board: Board;
  totRemoved: number;
}> => {
  const newBoard = b.filter((_row, ridx) => !lineIdxs.some(idx => idx === ridx));
  return {
    board: newBoard,
    totRemoved: b.length - newBoard.length
  };
};

export const mkRow = (len: number, b: Tile) => [...Array(len).fill(b)];
export const mkEmptyRow = mkRow(BOARD_CELLS, NoTetro);

export const appendEmptyRowsToBoard = (b: Board, amount: number): Board => [
  ...Array(amount).fill(mkEmptyRow),
  ...b
];

export const detectAndRemoveCompletedRows = (b: Board): Board =>
  pipe(
    getCompleteRowIdxs(b),
    idxsRowCompleted => removeCompleteRowFromBoard(b, idxsRowCompleted),
    ({ board, totRemoved }) => appendEmptyRowsToBoard(board, totRemoved)
  );
