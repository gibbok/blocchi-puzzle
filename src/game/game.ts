/**
 * Calculate the game score.
 * @param rowCompleted Number of rows completed
 */
export const calcScore = (rowCompleted: number): number => rowCompleted * 100;

/**
 * Calculate the game level.
 * @param score Game score
 */
export const calcLevel = (score: number): number => Math.ceil((score + 1) / 500);
