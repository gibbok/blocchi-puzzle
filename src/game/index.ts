export * from './board';
export * from './game';
export * from './interaction';
export * from './pieces';
export * from './tetromino';
export * from './types';
