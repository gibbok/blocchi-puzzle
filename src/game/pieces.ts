import { TetroPieces, Z, S, J, T, I, L, O } from './types';
export const pieces: TetroPieces = {
  Z: {
    N: [
      [Z, Z, 0],
      [0, Z, Z]
    ],
    E: [
      [0, Z],
      [Z, Z],
      [Z, 0]
    ],
    S: [
      [Z, Z, 0],
      [0, Z, Z]
    ],
    W: [
      [0, Z],
      [Z, Z],
      [Z, 0]
    ]
  },
  S: {
    N: [
      [0, S, S],
      [S, S, 0]
    ],
    E: [
      [S, 0],
      [S, S],
      [0, S]
    ],
    S: [
      [0, S, S],
      [S, S, 0]
    ],
    W: [
      [S, 0, 0],
      [S, S, 0],
      [0, S, 0]
    ]
  },
  J: {
    N: [
      [0, J],
      [0, J],
      [J, J]
    ],
    E: [
      [J, 0, 0],
      [J, J, J]
    ],
    S: [
      [J, J],
      [J, 0],
      [J, 0]
    ],
    W: [
      [J, J, J],
      [0, 0, J]
    ]
  },
  T: {
    N: [
      [T, T, T],
      [0, T, 0]
    ],
    E: [
      [0, T],
      [T, T],
      [0, T]
    ],
    S: [
      [0, T, 0],
      [T, T, T]
    ],
    W: [
      [T, 0],
      [T, T],
      [T, 0]
    ]
  },
  I: {
    N: [
      [0, I, 0, 0],
      [0, I, 0, 0],
      [0, I, 0, 0],
      [0, I, 0, 0]
    ],
    E: [
      [0, 0, 0, 0],
      [I, I, I, I],
      [0, 0, 0, 0],
      [0, 0, 0, 0]
    ],
    S: [
      [0, I, 0, 0],
      [0, I, 0, 0],
      [0, I, 0, 0],
      [0, I, 0, 0]
    ],
    W: [
      [0, 0, 0, 0],
      [I, I, I, I],
      [0, 0, 0, 0],
      [0, 0, 0, 0]
    ]
  },
  L: {
    N: [
      [L, 0],
      [L, 0],
      [L, L]
    ],
    E: [
      [L, L, L],
      [L, 0, 0]
    ],
    S: [
      [L, L],
      [0, L],
      [0, L]
    ],
    W: [
      [0, 0, L],
      [L, L, L]
    ]
  },
  O: {
    N: [
      [O, O],
      [O, O]
    ],
    E: [
      [O, O],
      [O, O]
    ],
    S: [
      [O, O],
      [O, O]
    ],
    W: [
      [O, O],
      [O, O]
    ]
  }
};
