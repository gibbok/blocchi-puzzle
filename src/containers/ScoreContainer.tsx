import { connect } from 'react-redux';
import { mkPublicState } from '../store/reducer';
import { Score } from '../components/Score';
import { InternalState } from '../game/types';

const mapStateToProps = (prevState: InternalState): { score: number; level: number } => {
  const { score, level } = mkPublicState(prevState);
  return { score, level };
};

export const ScoreContainer = connect(mapStateToProps)(Score);
