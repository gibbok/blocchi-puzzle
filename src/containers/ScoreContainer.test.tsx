import { mapStateToProps, ScoreContainer } from './ScoreContainer';
import { TetroEnum } from '../game/types';
import { mkInitialState } from '../store';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import React from 'react';
import { Provider } from 'react-redux';

describe('<ScoreContainer />', () => {
  it('should mapStateToProps correctly', () => {
    const input = mkInitialState(TetroEnum.I, TetroEnum.J);
    const output = mapStateToProps(input);
    const result = { score: 0, level: 1 };

    expect(output).toMatchObject(result);
  });

  it('should render correctly', () => {
    const mockStore = configureMockStore();
    const initialState = mkInitialState(TetroEnum.I, TetroEnum.J);
    const store = mockStore(initialState);
    const withProvider = (
      <Provider store={store}>
        <ScoreContainer />
      </Provider>
    );
    const component = renderer.create(withProvider);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
