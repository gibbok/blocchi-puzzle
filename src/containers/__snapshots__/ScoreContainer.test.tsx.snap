// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreContainer /> should render correctly 1`] = `
<div
  className="sc-AxhUy liFohR"
>
  <div
    className="sc-AxirZ jPhnSh"
  >
    <div
      className="sc-AxiKw bLKSBM"
    >
      Score
    </div>
    <div
      className="sc-AxiKw bLKSBM"
    >
      0
    </div>
  </div>
  <div
    className="sc-AxirZ jPhnSh"
  >
    <div
      className="sc-AxiKw bLKSBM"
    >
      Level
    </div>
    <div
      className="sc-AxiKw bLKSBM"
    >
      1
    </div>
  </div>
</div>
`;
